### PR TITLE
key backup: Implement setup screen #2883

### PR DESCRIPTION
**Pull request by BillCarsonFr**
initial commit

### Pull Request Checklist

* [ ] Changes has been tested on an Android device or Android emulator with API 16
* [ ] Pull request is based on the develop branch
* [ ] Pull request updates [CHANGES.rst](https://github.com/vector-im/riot-android/blob/develop/CHANGES.rst)
* [ ] Pull request includes screenshots or videos if containing UI changes
* [ ] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

File: vector/build.gradle [hunk 1]
```diff
@@ -238,6 +241,9 @@ dependencies {
     implementation 'com.github.bumptech.glide:glide:4.7.1'
     kapt 'com.github.bumptech.glide:compiler:4.7.1'
 
+    // Key backup helper
```
Comment (bmarty) [inline vector/build.gradle]:
```suggestion
    // Passphrase strength helper
```

File: vector/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -403,10 +399,14 @@
...
-
         <activity
-            android:name="im.vector.activity.MediaPreviewerActivity"
+            android:name=".activity.MediaPreviewerActivity"
             android:configChanges="orientation|screenSize" />
+        <activity
+            android:name=".activity.KeybackupSetupActivity"
+            android:configChanges="orientation|screenSize"
+            android:label="@string/title_activity_key_backup_setup"
+            android:windowSoftInputMode="stateHidden|adjustResize"></activity>
```
Comment (bmarty) [inline vector/src/main/AndroidManifest.xml]:
```suggestion
            android:windowSoftInputMode="stateHidden|adjustResize" />
```

File: vector/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -403,10 +399,14 @@
...
             android:label="@string/settings_notification_privacy" />
         <activity android:name=".activity.VectorWebViewActivity" />
-
         <activity
-            android:name="im.vector.activity.MediaPreviewerActivity"
+            android:name=".activity.MediaPreviewerActivity"
             android:configChanges="orientation|screenSize" />
+        <activity
+            android:name=".activity.KeybackupSetupActivity"
+            android:configChanges="orientation|screenSize"
```
Comment (bmarty) [inline vector/src/main/AndroidManifest.xml]: I do not like to add configChanges. Is it necessary?

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    override fun getMenuRes() = R.menu.keybackup_setup
+
+    override fun initUiAndData() {
+        if (isFirstCreation()) {
+            supportFragmentManager.beginTransaction()
+                    .replace(R.id.container, KeybackupSetupStep1Fragment.newInstance())
+                    .commitNow()
+        }
+        setSupportActionBar(findViewById(R.id.keybackup_toolbar))
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: Toolbar could have id `toolbar` and then you can just call `configureToolbar()`

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            return true
+        }
+        return super.onOptionsItemSelected(item)
+    }
+
+    private fun onSkip() {
+        AlertDialog.Builder(this)
+                .setTitle(R.string.keybackup_setup_skip_title)
+                .setMessage(R.string.keybackup_setup_skip_msg)
+                .setNegativeButton(R.string.cancel) { dialog, id ->
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: You can use shorter form `setNegativeButton(R.string.cancel, null)`

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                    //nop
+                }
+                .setPositiveButton(R.string.skip) { dialog, id ->
+                    setResult(Activity.RESULT_CANCELED)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: Not necessary, RESULT_CANCELED is the default value

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import android.arch.lifecycle.MutableLiveData
+import android.arch.lifecycle.ViewModel
+import android.text.TextUtils
+import com.nulabinc.zxcvbn.Strength
+import org.matrix.androidsdk.crypto.keysbackup.MegolmBackupCreationInfo
+
+/**
+ * The shared view model between all fragments.
+ */
+class KeybackupSetupSharedViewModel  : ViewModel() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt]: Please format this file

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        isCreatingBackupVersion.value = false
+    }
+
+    fun checkValidity(): Boolean {
+        if (TextUtils.isEmpty(passphrase.value)) {
+            return false
+        }
+        if (TextUtils.isEmpty(confirmPassphrase.value)) {
+            return false
+        } else if (confirmPassphrase.value != passphrase.value) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt]: `else` is not necessary

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+        addSource(passwordStrength) { value = checkValidity() }
+    }
+
+    init {
+        showPasswordMode.value = false
+        recoveryKey.value = null
+        isCreatingBackupVersion.value = false
+    }
+
+    fun checkValidity(): Boolean {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt]: should be named `isValid` as it returns a boolean

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    @BindView(R.id.keybackupsetup_step1_button)
+    lateinit var mButton: Button
+
+    companion object {
+        fun newInstance() = KeybackupSetupStep1Fragment()
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        mButton.setOnClickListener() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt]: Use Butterknife @OnClick annotation

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            val fm = this.activity?.supportFragmentManager
+            fm?.beginTransaction()?.replace(R.id.container, KeybackupSetupStep2Fragment.newInstance())?.addToBackStack(null)?.commit()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt]: I prefer to see one call per line, for code clarity

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+                mRecoveryKeyTextView.text = newValue
+                mCopyButton.visibility = View.VISIBLE
+                mFinishButton.visibility = View.VISIBLE
+            }
+        })
+
+        mCopyButton.setOnClickListener {
+            val recoveryKey = viewModel.recoveryKey.value
+            if (recoveryKey != null) {
+                val share = Intent(android.content.Intent.ACTION_SEND)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt]: Move to SystemUtils?

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+                // what to do with it.
+                share.putExtra(Intent.EXTRA_SUBJECT, context?.getString(R.string.recovery_key))
+                share.putExtra(Intent.EXTRA_TEXT, recoveryKey)
+
+                startActivity(Intent.createChooser(share, context?.getString(R.string.keybackup_setup_step3_share_intent_chooser_title)))
+                viewModel.copyHasBeenMade = true
+            }
+        }
+
+        mFinishButton.setOnClickListener {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt]: Use @OnClick

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                            viewModel.isCreatingBackupVersion.value = false
+                            activity?.setResult(Activity.RESULT_OK)
+                            activity?.finish()
+                        }
+
+                        override fun onUnexpectedError(e: Exception?) {
+                            Log.e(LOG_TAG, "## createKeyBackupVersion ${e?.localizedMessage}")
+                            viewModel.isCreatingBackupVersion.value = false
+                            activity?.run {
+                                runOnUiThread {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt]: runOnUiThread should not be necessary

Comment (bmarty) [reply vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep3Fragment.kt]: (the others neither)

File: vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+/*
+A password strength bar custom widget
+ Strength is an Integer
+ # -1 No strength
+ # 0 Weak
+ # 1 Fair
+ # 2 Good
+ # 3 Strong
+ */
+class PasswordStrengthBar(context: Context, attrs: AttributeSet? = null) : LinearLayout(context, attrs) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt]: You should use `@JvmOverload` for the constructor

File: vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        LayoutInflater.from(context)
+                .inflate(R.layout.password_strength_bar, this, true)
+        orientation = HORIZONTAL
+        ButterKnife.bind(this)
+        strength = -1
+    }
+
+    companion object {
+        private val colors = listOf<Int>(
+                Color.parseColor("#f56679"),
```
Comment (bmarty) [inline vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt]: This color should be defined as resources. You can then bind them with `@BindColor`

File: vector/src/main/res/layout/keybackup_setup_activity.xml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+    <FrameLayout
+        android:id="@+id/container"
+        android:layout_width="0dp"
+        android:layout_height="0dp"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/keybackup_toolbar"
+        tools:context="im.vector.activity.KeybackupSetupActivity" />
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_activity.xml]: It's strange to have a context here. Should be in root view

File: vector/src/main/res/layout/keybackup_setup_fragment.xml [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+        android:background="?attr/colorAccent"
+        android:minWidth="200dp"
+        android:padding="8dp"
+        android:text="@string/keybackup_setup_step1_button_title"
+        android:textColor="@android:color/white"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/keybackupsetup_step1_description"
+
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_fragment.xml]: remove extra new line

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+            android:layout_marginEnd="20dp"
+            android:text="@string/keybackup_setup_step2_text"
+            android:textAlignment="center"
+            app:layout_constraintBottom_toTopOf="@id/keybackup_passphrase_container"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
+
+
+        <android.support.constraint.ConstraintLayout
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: This ConstraintLayout power is not really used... You could try to flatten view hierarchy

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+                            android:id="@+id/keybackup_passphrase_enter_edittext"
+                            android:layout_width="match_parent"
+                            android:layout_height="wrap_content"
+                            android:hint="@string/keybackup_setup_step2_passphrase_enter_hint"
+                            android:maxLines="3"
+                            android:textColor="?android:textColorPrimary" />
+
+                    </android.support.design.widget.TextInputLayout>
+
+                    <RelativeLayout
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: This RelativeLayout should be removed

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+                    android:layout_marginStart="16dp"
+                    android:layout_marginLeft="16dp"
+                    android:layout_marginEnd="44dp"
+                    android:layout_marginRight="44dp" />
+
+                <View
+                    android:layout_width="match_parent"
+                    android:layout_height="1dp"
+                    android:layout_margin="6dp"
+                    android:background="@color/list_divider_color_light" />
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: You must use `?attr` and not directly color, to correctly support all the themes

File: vector/src/main/res/layout/keybackup_setup_step3_fragment.xml [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="20dp"
+        android:layout_marginBottom="20dp"
+        app:layout_constraintBottom_toTopOf="@id/keybackupsetup_step3_copy_button"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/keybackupsetup_step3_text">
+
+        <LinearLayout
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_step3_fragment.xml]: To remove, use ConstraintLayout capabilities, or remove the ConstraintLayout

File: vector/src/main/res/layout/keybackup_setup_step3_fragment.xml [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+                android:layout_height="wrap_content"
+                android:text="@string/recovery_key"
+                android:textAlignment="center"
+                android:textStyle="bold" />
+
+            <View
+                android:layout_width="match_parent"
+                android:layout_height="1dp"
+                android:layout_margin="6dp"
+                android:background="@color/list_divider_color_light" />
```
Comment (bmarty) [inline vector/src/main/res/layout/keybackup_setup_step3_fragment.xml]: use ?attr

File: vector/src/main/res/layout/password_strength_bar.xml [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+    tools:parentTag="android.widget.LinearLayout">
+
+    <View
+        android:id="@+id/password_strength_bar_1"
+        android:layout_width="0dp"
+        android:layout_height="4dp"
+        android:layout_weight="1"
+        android:background="#f56679" />
+
+    <View
```
Comment (bmarty) [inline vector/src/main/res/layout/password_strength_bar.xml]: Remove intermediate View like this and use margin on other Views

File: vector/src/main/res/layout/password_strength_bar.xml [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+        android:layout_width="5dp"
+        android:layout_height="4dp" />
+
+    <View
+        android:id="@+id/password_strength_bar_2"
+        android:layout_width="0dp"
+        android:layout_height="4dp"
+        android:layout_weight="1"
+        android:background="#ffc666" />
```
Comment (bmarty) [inline vector/src/main/res/layout/password_strength_bar.xml]: should be tools:background

File: vector/src/main/res/menu/keybackup_setup.xml [hunk 26]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<menu xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <item
+        android:id="@+id/ic_action_keybackup_setup_skip"
+        android:title="@string/auth_skip"
```
Comment (bmarty) [inline vector/src/main/res/menu/keybackup_setup.xml]: Use `@string/skip`

File: vector/src/main/AndroidManifest.xml (hunk 3 above)
Comment (BillCarsonFr) [reply vector/src/main/AndroidManifest.xml]: removed, was just copy/paste

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt (hunk 6 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: ok

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt (hunk 5 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: done

File: vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt (hunk 4 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/activity/KeybackupSetupActivity.kt]: ok, changed to toolbar

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt (hunk 7 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt]: done

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt (hunk 8 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupSharedViewModel.kt]: code removed (not used anymore as the button is always enabled)

File: vector/src/main/res/layout/keybackup_setup_step3_fragment.xml (hunk 23 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_step3_fragment.xml]: done

File: vector/src/main/res/layout/password_strength_bar.xml (hunk 24 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/password_strength_bar.xml]: done

File: vector/src/main/res/layout/password_strength_bar.xml (hunk 25 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/password_strength_bar.xml]: done

File: vector/src/main/res/menu/keybackup_setup.xml (hunk 26 above)
Comment (BillCarsonFr) [reply vector/src/main/res/menu/keybackup_setup.xml]: done

File: vector/src/main/res/layout/keybackup_setup_step3_fragment.xml (hunk 22 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_step3_fragment.xml]: done (use power)

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt (hunk 10 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt]: done

File: vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt (hunk 11 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/fragments/keybackupsetup/KeybackupSetupStep1Fragment.kt]: done

File: vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt (hunk 15 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt]: done

File: vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt (hunk 16 above)
Comment (BillCarsonFr) [reply vector/src/main/java/im/vector/ui/PasswordStrengthBar.kt]: done

File: vector/src/main/res/layout/keybackup_setup_activity.xml (hunk 17 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_activity.xml]: moved up

File: vector/src/main/res/layout/keybackup_setup_fragment.xml (hunk 18 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_fragment.xml]: done

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml (hunk 19 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: done

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml (hunk 20 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: it's the clickable area, what do you recommand?

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml (hunk 21 above)
Comment (BillCarsonFr) [reply vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: done

File: vector/src/main/res/layout/keybackup_setup_step2_fragment.xml (hunk 20 above)
Comment (bmarty) [reply vector/src/main/res/layout/keybackup_setup_step2_fragment.xml]: The image could be clickable

**Pull request merged**